cpp/include/raft/neighbors/detail/knn_brute_force.cuh

```diff
@@ -0,0 +1,445 @@
+
+    out_D = all_D.data();
+    out_I = all_I.data();
+  }
+
+  // currently we don't support col_major inside tiled_brute_force_knn, because
+  // of limitattions of the pairwise_distance API:
+  // 1) paiwise_distance takes a single 'isRowMajor' parameter - and we have
+  // multiple options here (like rowMajorQuery/rowMajorIndex)
+  // 2) because of tiling, we need to be able to set a custom stride in the PW
+  // api, which isn't supported
+  // Instead, transpose the input matrices if they are passed as col-major.
+  auto search = search_items;
+  rmm::device_uvector<value_t> search_row_major(0, userStream);
+  if (!rowMajorQuery) {
+    search_row_major.resize(n * D, userStream);
+    raft::linalg::transpose(handle, search, search_row_major.data(), n, D, userStream);
+    search = search_row_major.data();
+  }
+
+  // Make other streams from pool wait on main stream
+  handle.wait_stream_pool_on_stream();
+
+  for (size_t i = 0; i < input.size(); i++) {
+    value_t* out_d_ptr = out_D + (i * k * n);
+    IdxType* out_i_ptr = out_I + (i * k * n);
+
+    auto stream = handle.get_next_usable_stream(i);
+
+    if (k <= 64 && rowMajorQuery == rowMajorIndex && rowMajorQuery == true &&
+        (metric == raft::distance::DistanceType::L2Unexpanded ||
+         metric == raft::distance::DistanceType::L2SqrtUnexpanded ||
+         metric == raft::distance::DistanceType::L2Expanded ||
+         metric == raft::distance::DistanceType::L2SqrtExpanded)) {
+      fusedL2Knn(D,
+                 out_i_ptr,
+                 out_d_ptr,
+                 input[i],
+                 search_items,
+                 sizes[i],
+                 n,
+                 k,
+                 rowMajorIndex,
+                 rowMajorQuery,
+                 stream,
+                 metric);
+
+      // Perform necessary post-processing
+      if (metric == raft::distance::DistanceType::L2SqrtExpanded ||
+          metric == raft::distance::DistanceType::L2SqrtUnexpanded ||
+          metric == raft::distance::DistanceType::LpUnexpanded) {
+        float p = 0.5;  // standard l2
+        if (metric == raft::distance::DistanceType::LpUnexpanded) p = 1.0 / metricArg;
+        raft::linalg::unaryOp<float>(
+          res_D,
+          res_D,
+          n * k,
+          [p] __device__(float input) { return powf(fabsf(input), p); },
+          stream);
+      }
+    } else {
+      switch (metric) {
+        case raft::distance::DistanceType::Haversine:
+          ASSERT(D == 2,
+                 "Haversine distance requires 2 dimensions "
+                 "(latitude / longitude).");
+
+          haversine_knn(out_i_ptr, out_d_ptr, input[i], search_items, sizes[i], n, k, stream);
+          break;
+        default:
+          // Create a new handle with the current stream from the stream pool
+          raft::device_resources stream_pool_handle(handle);
+          raft::resource::set_cuda_stream(stream_pool_handle, stream);
```

Comment:
I'm going to add a copy overload to the handle that accepts a stream to we can make this easier.